File: cs50/code/week1-c/float-bits.c
#include <stdio.h>
#include <stddef.h> // Required for size_t

// Function to print the bits of a single byte (remains the same)
void print_byte_bits(unsigned char byte)
{
    // Iterate from the most significant bit (MSB) to the least (LSB)
    for (int i = 7; i >= 0; i--)
    {
        // Shift the byte right by 'i' positions
        // Use bitwise AND with 1 to isolate the bit at the current position
        int bit = (byte >> i) & 1;
        printf("%d", bit);
    }
}

int main()
{
    // Use 'double' for standard 64-bit floating point
    double my_double = 0.1; // Example double value (e)

    // Use %.17g for printing doubles to see full precision generally
    printf("Double value: %.17g\n", my_double);
    printf("Size of double: %zu bytes (%zu bits)\n", sizeof(my_double), sizeof(my_double) * 8);

    // Get a pointer to the raw bytes of the double
    // unsigned char is used because it's typically 1 byte
    unsigned char *byte_ptr = (unsigned char *)&my_double;

    printf("Memory representation (bits, byte by byte in memory order):\n");

    // Iterate through each byte of the double (usually 8 bytes)
    // The order depends on the system's endianness
    for (size_t i = 0; i < sizeof(my_double); i++)
    {
        printf("Byte %zu: ", i);
        print_byte_bits(byte_ptr[i]);
        printf("  (Hex: 0x%02X)\n", byte_ptr[i]); // Also show hex value
    }
    printf("\n");

    printf("Combined bits (memory order): ");
    for (size_t i = 0; i < sizeof(my_double); i++)
    {
        print_byte_bits(byte_ptr[i]);
        printf(" "); // Space between bytes
    }
    printf("\n\n");

    printf("--- Endianness Note ---\n");
    printf("The order of the bytes shown above is the order they appear in memory.\n");
    printf(" - On Little-Endian systems (common, e.g., x86), the least significant byte is stored first (at the lowest memory address, Byte 0).\n");
    printf(" - On Big-Endian systems, the most significant byte is stored first.\n");
    printf("The IEEE 754 double-precision standard defines a logical order (1 Sign | 11 Exponent | 52 Mantissa), but the storage order in memory might differ.\n");

    return 0;
}

File: cs50/code/week1-c/float-bits.py
import struct
import sys

# The Python float you want to inspect
my_float = 0.1

# Pack the Python float into bytes.
# 'd' is the format code for a C double (which Python floats usually are).
# By default, struct.pack uses the machine's native byte order (endianness).
try:
    # Use 'd' for native C double format and endianness
    packed_bytes = struct.pack("d", my_float)
    # Alternatively, force endianness if needed:
    # packed_bytes = struct.pack('<d', my_float) # Little-endian
    # packed_bytes = struct.pack('>d', my_float) # Big-endian
except struct.error as e:
    print(f"Error packing float: {e}")
    print(
        "This might happen on unusual platforms where 'float' isn't a standard double."
    )
    exit()

print(f"Python float value: {my_float}")
print(
    f"Underlying C Type Size: {struct.calcsize('d')} bytes ({struct.calcsize('d') * 8} bits)"
)
print(f"System Endianness: {sys.byteorder}")
print("-" * 30)

print("Memory Representation (Bytes in native storage order):")

bit_representations_memory_order = []

# Iterate through the bytes object returned by struct.pack
for i, byte_val in enumerate(packed_bytes):
    # byte_val is an integer representing the byte's value (0-255)
    # Format the integer as an 8-digit binary string, padded with leading zeros
    bits = format(byte_val, "08b")
    hex_val = format(byte_val, "02X")  # Also show Hex
    print(f"Byte {i}: {bits}  (Hex: 0x{hex_val})")
    bit_representations_memory_order.append(bits)

print("\nCombined bits (native memory storage order):")
print(" ".join(bit_representations_memory_order))
print("-" * 30)

# --- Deconstruct based on IEEE 754 (assuming standard double) ---

# Reconstruct the logical bit pattern (Sign | Exponent | Mantissa)
# This depends on the system's endianness used during packing

if sys.byteorder == "little":
    # On little-endian, struct.pack puts LSB first, so reverse bytes for logical view
    logical_bit_string = "".join(reversed(bit_representations_memory_order))
    print("Reconstructed logical IEEE 754 order (Little-Endian System):")
elif sys.byteorder == "big":
    # On big-endian, struct.pack puts MSB first, matching logical view
    logical_bit_string = "".join(bit_representations_memory_order)
    print("Logical IEEE 754 order (Big-Endian System):")
else:
    logical_bit_string = None
    print("Could not determine logical order for unknown endianness.")

if logical_bit_string:
    # Extract S, E, F from the 64-bit string
    sign_bit = logical_bit_string[0]
    exponent_bits = logical_bit_string[1:12]  # 11 bits
    fraction_bits = logical_bit_string[12:]  # 52 bits

    # Calculate exponent value
    exponent_value = int(exponent_bits, 2)
    actual_exponent = exponent_value - 1023  # Bias for double precision

    print(f"  Sign Bit: {sign_bit} ({'Positive' if sign_bit == '0' else 'Negative'})")
    print(
        f"  Exponent Bits: {exponent_bits} (Stored Value: {exponent_value}, Actual Exponent: {actual_exponent})"
    )
    print(f"  Fraction Bits: {fraction_bits}")
